packages/material-ui-lab/src/Masonry/Masonry.js

```diff
@@ -0,0 +1,111 @@
+// A masonry component using the following libs as inspiration.
+// - https://github.com/STRML/react-grid-layout
+// - https://next.material-ui.com/components/image-list/#masonry-image-list
+// - https://github.com/desandro/masonry
+// - https://w3bits.com/css-grid-masonry
+
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
+import { unstable_composeClasses as composeClasses } from '@material-ui/unstyled';
+import { styled, useThemeProps } from '@material-ui/core/styles';
+import { getMasonryUtilityClass } from './masonryClasses';
+import MasonryContext from './MasonryContext';
+
+const useUtilityClasses = (styleProps) => {
+  const { classes } = styleProps;
+
+  const slots = {
+    root: ['root'],
+  };
+
+  return composeClasses(slots, getMasonryUtilityClass, classes);
+};
+
+const MasonryRoot = styled('div', {
+  name: 'MuiMasonry',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    return [styles.root];
+  },
+})(({ styleProps, theme }) => {
+  /* Styles applied to the root element. */
+  return {
+    display: 'grid',
+    gridAutoRows: 0,
+    padding: 0,
+    overflow: 'auto',
+    width: '100%',
+    columnGap: theme.spacing(styleProps.spacing),
```

Comment:
We need to handle the breakpoints values here, they can be defined as arrays or objects. See https://next.material-ui.com/system/basics/#responsive-values You can take a look on how the `Stack` components solves this. Also, would suggest doing the same for the `cols` property if possible.

This should work - https://codesandbox.io/s/imagemasonry-material-demo-forked-5ogd5?file=/demo.js